### PR TITLE
Download custom clang-tidy from s3

### PR DESCRIPTION
**Pull request by 1ntEgr8**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61121
* **#61120**
* #61119

This change implements downloading the clang-tidy binary from S3. It
does this by lifting the existing implementation of the download routine
(which was in `tools/linter/clang_format_utils.py`) to a generic
`tools/linter/install/download_bin.py`.

Folder organization:
```
tools/linter/install
|_ clang_tidy.py
|_ download_bin.py
|_ hashes/
|_ __init__.py

```

`download_bin.py` contains the code for downloading and verifying a
binary from specified sources.  `clang_tidy.py` declares the inputs for
installing `clang-tidy` `hashes/` contains sha256sum hashes of the
binaries (used for verification)

The command to download a linter binary will look like this:
```
python3 tools/linter/install/<binary-name>.py
```

For example, to install `clang-tidy`:
```
python3 tools/linter/install/clang_tidy.py
```

Based on the inputs declared in `tools/linter/install/clang_tidy.py`,
this will install our custom build of `clang-tidy` at
`$PYTORCH_ROOT/.clang-tidy-bin/clang-tidy`

Test plan:

Successful download:
```
python3 tools/linter/install/clang_tidy.py
.clang-tidy-bin/clang-tidy --checks="*" --list-checks | grep "misc-max-tokens"
```

Unsuccessful download:
```
echo "bad hash" > tools/linter/install/hashes/clang-tidy-linux64
python3 tools/linter/install/clang_tidy.py # this should fail
```

### CODE REVIEW

File: mypy.ini [hunk 1]
```diff
@@ -39,6 +39,7 @@ files =
     test/test_utils.py,
     tools/linter/clang_format_utils.py,
     tools/linter/clang_tidy,
+    tools/linter/install,
     tools/generate_torch_version.py,
     tools/render_junit.py,
     tools/stats
```
Comment (samestep) [inline mypy.ini:45]: same comment here as https://github.com/pytorch/pytorch/pull/61119#discussion_r662401301

File: tools/linter/clang_format_utils.py [hunk 2]
```diff
@@ -1,16 +1,5 @@
...
 import os
-import urllib.request
-import urllib.error
-
-# String representing the host platform (e.g. Linux, Darwin).
-HOST_PLATFORM = platform.system()
-
-# PyTorch directory root, derived from the location of this file.
-PYTORCH_ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.realpath(__file__))))
+from install.download_bin import download, PYTORCH_ROOT  # type: ignore[import]
```
Comment (samestep) [inline tools/linter/clang_format_utils.py:2]: why is the `# type: ignore[import]` needed here?

File: tools/linter/install/clang_tidy.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import os
+from download_bin import download, PYTORCH_ROOT, HASH_PATH  # type: ignore[import]
```
Comment (samestep) [inline tools/linter/install/clang_tidy.py:2]: same question about `# type: ignore[import]` here

File: tools/linter/install/hashes/clang-tidy-linux64 [hunk 4]
```diff
@@ -0,0 +1 @@
+44103f662c866921dabd40d047788935c43402d260be10613ab2111e01726892
```
Comment (janeyx99) [inline tools/linter/install/hashes/clang-tidy-linux64:1]: Is this file supposed to be committed vs in the .gitignore?

Comment (samestep) [reply tools/linter/install/hashes/clang-tidy-linux64:1]: yeah it's similar to these two existing files:

https://github.com/pytorch/pytorch/blob/80cab10534e1a9ee6d2493113f169a2570979651/tools/clang_format_hash/linux64/clang-format-linux64#L1

https://github.com/pytorch/pytorch/blob/80cab10534e1a9ee6d2493113f169a2570979651/tools/clang_format_hash/mac/clang-format-mojave#L1

Comment (1ntEgr8) [reply tools/linter/install/hashes/clang-tidy-linux64:1]: Yes, we need to commit the hash since its used for verifying the binary.

**Pull request closed without merging**